esp-hal/src/dma/m2m.rs

```diff
@@ -31,145 +40,467 @@ pub struct Mem2Mem<'d, Dm>
 where
     Dm: DriverMode,
 {
-    channel: Channel<Dm, AnyGdmaChannel<'d>>,
-    rx_chain: DescriptorChain,
-    tx_chain: DescriptorChain,
-    peripheral: DmaPeripheral,
+    /// RX Half
+    pub rx: Mem2MemRx<'d, Dm>,
+    /// TX Half
+    pub tx: Mem2MemTx<'d, Dm>,
 }
 
 impl<'d> Mem2Mem<'d, Blocking> {
     /// Create a new Mem2Mem instance.
     pub fn new(
         channel: impl DmaChannelConvert<AnyGdmaChannel<'d>>,
         peripheral: impl DmaEligible,
-        rx_descriptors: &'static mut [DmaDescriptor],
-        tx_descriptors: &'static mut [DmaDescriptor],
-    ) -> Result<Self, DmaError> {
-        unsafe {
-            Self::new_unsafe(
-                channel,
-                peripheral.dma_peripheral(),
-                rx_descriptors,
-                tx_descriptors,
-                crate::dma::CHUNK_SIZE,
-            )
-        }
-    }
-
-    /// Create a new Mem2Mem instance with specific chunk size.
-    pub fn new_with_chunk_size(
-        channel: impl DmaChannelConvert<AnyGdmaChannel<'d>>,
-        peripheral: impl DmaEligible,
-        rx_descriptors: &'static mut [DmaDescriptor],
-        tx_descriptors: &'static mut [DmaDescriptor],
-        chunk_size: usize,
-    ) -> Result<Self, DmaError> {
-        unsafe {
-            Self::new_unsafe(
-                channel,
-                peripheral.dma_peripheral(),
-                rx_descriptors,
-                tx_descriptors,
-                chunk_size,
-            )
-        }
+    ) -> Self {
+        unsafe { Self::new_unsafe(channel, peripheral.dma_peripheral()) }
     }
 
     /// Create a new Mem2Mem instance.
     ///
     /// # Safety
     ///
-    /// You must ensure that your not using DMA for the same peripheral and
-    /// that your the only one using the DmaPeripheral.
+    /// You must ensure that you're not using DMA for the same peripheral and
+    /// that you're the only one using the DmaPeripheral.
     pub unsafe fn new_unsafe(
         channel: impl DmaChannelConvert<AnyGdmaChannel<'d>>,
         peripheral: DmaPeripheral,
+    ) -> Self {
+        let mut channel = Channel::new(channel.degrade());
+
+        channel.rx.set_mem2mem_mode(true);
+
+        Mem2Mem {
+            rx: Mem2MemRx {
+                channel: channel.rx,
+                peripheral,
+            },
+            tx: Mem2MemTx {
+                channel: channel.tx,
+                peripheral,
+            },
+        }
+    }
+
+    /// Shortcut to create a [SimpleMem2Mem]
+    pub fn with_descriptors(
+        self,
         rx_descriptors: &'static mut [DmaDescriptor],
         tx_descriptors: &'static mut [DmaDescriptor],
```

Comment:
Do you think we could maybe get away with `&'d mut [DmaDescriptor]`?

Reply:
With more `unsafe` yeah can do.
The unsafe is because `SimpleMem2Mem` might be forgotten, and the user might free the descriptors whilst they're still in use.

Reply:
I think we pretty much established that we consider forget "unsafe", and assume it doesn't exist for soundness questions like this.

Reply:
That's fair enough. I'll still need it to construct the DMA buffers, but I can say that as the safety statement.